src/engine/guidance/post_processing.cpp

```diff
@@ -1037,6 +1039,9 @@ std::vector<RouteStep> collapseTurns(std::vector<RouteStep> steps)
     // a series of turns is only possible to collapse if its only name changes and suppressed turns.
     const auto canCollapseAll = [&steps](std::size_t index, const std::size_t end_index) {
         BOOST_ASSERT(end_index <= steps.size());
+        if (!compatible(steps[index], steps[index + 1]))
+            return false;
+        ++index;
```

Comment:
The respective call site: https://github.com/Project-OSRM/osrm-backend/pull/3246/files/31ae8492dbecf266d8d4413348f03805ac621ef1#diff-99adcb8bccf8e94a176bdf92509ff070L1139